src/main/java/it/gov/pagopa/reward/service/synchronous/RewardTrxSynchronousApiServiceImpl.java

```diff
@@ -0,0 +1,81 @@
+        this.initiativesEvaluatorFacadeService = initiativesEvaluatorFacadeService;
+        this.userInitiativeCountersRepository = userInitiativeCountersRepository;
+        this.trxPreviewRequest2TransactionDtoMapper = trxPreviewRequest2TransactionDtoMapper;
+        this.rewardTransaction2SynchronousTransactionRequestDTOMapper = rewardTransaction2SynchronousTransactionRequestDTOMapper;
+    }
+
+    @Override
+    public Mono<SynchronousTransactionResponseDTO> postTransactionPreview(SynchronousTransactionRequestDTO trxPreviewRequest, String initiativeId) {
+        log.trace("[REWARD] Starting reward preview calculation for transaction {}", trxPreviewRequest.getTransactionId());
+        TransactionDTO trxDTO = trxPreviewRequest2TransactionDtoMapper.apply(trxPreviewRequest);
+
+        return checkInitiative(trxPreviewRequest, initiativeId)
+                .flatMap(b -> checkOnboarded(trxPreviewRequest, trxDTO, initiativeId))
+                .flatMap(b -> userInitiativeCountersRepository.findById(UserInitiativeCounters.buildId(trxDTO.getUserId(),initiativeId)))
+                .map(userInitiativeCounters -> {
+                    UserInitiativeCountersWrapper counterWrapper = UserInitiativeCountersWrapper.builder()
+                            .userId(userInitiativeCounters.getUserId())
+                            .initiatives(Map.of(initiativeId, userInitiativeCounters))
+                            .build();
+                    return initiativesEvaluatorFacadeService.evaluateInitiativesBudgetAndRules(trxDTO, List.of(initiativeId), counterWrapper);
+                })
+                .map(p -> rewardTransaction2SynchronousTransactionRequestDTOMapper.apply(trxPreviewRequest.getTransactionId(),initiativeId, p.getSecond()));
+    }
+
+    private Mono<Boolean> checkInitiative(SynchronousTransactionRequestDTO request, String initiativeId){
+        return rewardContextHolderService.getInitiativeConfig(initiativeId).hasElement()
+                .map(b ->{
+                    if ( b.equals(Boolean.TRUE)){
+                        return Boolean.TRUE;
+                    } else {
+                        throw new TransactionSynchronousException(trxPreviewRequest2TransactionDtoMapper.apply(request, initiativeId, List.of(RewardConstants.TRX_REJECTION_REASON_INITIATIVE_NOT_FOUND)));
+                    }
+                });
+    }
+    private Mono<Boolean> checkOnboarded(SynchronousTransactionRequestDTO request, TransactionDTO trx, String initiativeId){
+        return onboardedInitiativesService.isOnboarded(trx.getHpan(),trx.getTrxDate(), initiativeId)
+                .map(b -> {
+                    if ( b.equals(Boolean.TRUE)){
+                        return Boolean.TRUE;
+                    } else {
+                        throw new TransactionSynchronousException(trxPreviewRequest2TransactionDtoMapper.apply(request, initiativeId, List.of(RewardConstants.TRX_REJECTION_REASON_NO_INITIATIVE)));
```

Comment:
as you can see, you almost rewriting what you have written in row 67

you can refactor extracting this "new" invocation